frontend/src/metabase/visualizations/echarts/pie/model/index.ts

```diff
@@ -123,108 +258,286 @@ export function getPieChartModel(
+        value,
         displayValue: value,
         normalizedPercentage: visible ? value / total : 0, // slice percentage values are normalized to 0-1 scale
-        rowIndex: rowIndiciesByKey.get(key),
-        color,
+        color: getColorForRing(
+          color,
+          "inner",
+          colDescs.middleDimensionDesc != null,
+          renderingContext,
+        ),
         visible,
+        children: new Map(),
+        column: colDescs.dimensionDesc.column,
+        rowIndex: checkNotNull(rowIndiciesByKey.get(key)),
+        legendHoverIndex: index,
         isOther,
-        noHover: false,
         includeInLegend: true,
+        startAngle: 0, // placeholders
+        endAngle: 0,
       };
     })
-    .filter(slice => isNonPositive || slice.value > 0)
+    .filter(slice => slice.value > 0)
     .partition(slice => slice != null && !slice.isOther)
     .value();
 
   // We don't show the grey other slice if there isn't more than one slice to
   // group into it
   if (others.length === 1) {
     const singleOtherSlice = others.pop();
-    slices.push(checkNotNull(singleOtherSlice));
+    sliceTreeNodes.push(checkNotNull(singleOtherSlice));
   }
 
+  sliceTreeNodes.forEach(node => {
+    // Map key needs to be string, because we use it for lookup with values from
+    // echarts, and echarts casts numbers to strings
+    sliceTree.set(String(node.key), node);
+  });
+
+  // Iterate through non-aggregated rows from query result to build layers for
+  // the middle and outer ring slices.
+  if (colDescs.middleDimensionDesc != null) {
+    const formatMiddleDimensionValue = getDimensionFormatter(
+      settings,
+      colDescs.middleDimensionDesc.column,
+      renderingContext.formatValue,
+    );
+
+    const formatOuterDimensionValue =
+      colDescs.outerDimensionDesc?.column != null
+        ? getDimensionFormatter(
+            settings,
+            colDescs.outerDimensionDesc.column,
+            renderingContext.formatValue,
+          )
+        : undefined;
+
+    dataRows.forEach((row, index) => {
+      // Needed to tell typescript it's defined
+      if (colDescs.middleDimensionDesc == null) {
+        throw new Error(`Missing middleDimensionDesc`);
+      }
+
+      const dimensionKey = getKeyFromDimensionValue(
+        row[colDescs.dimensionDesc.index],
+      );
+      const dimensionNode = sliceTree.get(String(dimensionKey));
+      const dimensionIsOther = dimensionNode == null;
+      if (dimensionIsOther) {
+        return;
+      }
+      const metricValue = getNumberOr(row[colDescs.metricDesc.index], 0);
+      if (metricValue < 0) {
+        return;
+      }
+
+      // Create or update node for middle dimension
+      const middleDimensionKey = getKeyFromDimensionValue(
+        row[colDescs.middleDimensionDesc.index],
+      );
+      let middleDimensionNode = dimensionNode.children.get(
+        String(middleDimensionKey),
+      );
+
+      if (middleDimensionNode == null) {
```

Comment:
This middle/outer dimension node logic looks quite similar, would it be possible to extract a function for node creation with shared logic?

Reply:
Yup, just did that 8dc70e4e8bbf81f4aee936ee6d18d236b3d63a31